File: neetcode/6. linked list/18.rotateList.cpp
#include <limits.h>

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(NULL) {}
  ListNode(int x) : val(x), next(NULL) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *rotateRight(ListNode *head, int k) {
    if (!head || !head->next || k == 0) return head;

    ListNode *tail = head;
    int len = 1;
    while (tail->next) {
      tail = tail->next;
      len++;
    }

    k = k % len;
    if (k == 0) return head;

    ListNode *newTail = head;
    for (int i = 0; i < len - k - 1; i++) {
      newTail = newTail->next;
    }

    ListNode *newHead = newTail->next;
    newTail->next = NULL;
    tail->next = head;

    return newHead;
  }
};

File: neetcode/6. linked list/18.rotateList.py
# Given the head of a linked list, rotate the list to the right by k places.

# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next

class Solution:
  def rotateRight(self, head: ListNode, k: int) -> ListNode:
    if not head or not head.next or k == 0:
      return head

    # get the length of the list
    length = 1
    tail = head
    while tail.next:
      length += 1
      tail = tail.next

    # get the actual number of rotations
    k = k % length

    # if k is 0, then no need to rotate
    if k == 0:
      return head

    # get the node right before the new head
    new_tail = head
    for _ in range(length - k - 1):
      new_tail = new_tail.next

    # get the new head
    new_head = new_tail.next

    # break the list
    new_tail.next = None
    tail.next = head

    return new_head